icu4c/source/common/unicode/utfiterator.h

```diff
@@ -0,0 +1,2502 @@
+                std::forward_iterator_tag,
+                typename std::iterator_traits<UnitIter>::iterator_category>;
+        CP32 c = *p;
+        ++p;
+        if constexpr (isMultiPass) {
+            return {c, 1, p0, p};
+        } else {
+            return {c, 1};
+        }
+    }
+
+    U_FORCE_INLINE static UnsafeCodeUnits<CP32, UnitIter> decAndRead(UnitIter &p) {
+        UnitIter p0 = p;
+        CP32 c = *--p;
+        return {c, 1, p, p0};
+    }
+};
+
+#endif
+
+// Validating iterators ---------------------------------------------------- ***
+
+/**
+ * Validating iterator over the code points in a Unicode string.
+ *
+ * The UnitIter can be
+ * an input_iterator, a forward_iterator, or a bidirectional_iterator (including a pointer).
+ * The UTFIterator will have the corresponding iterator_category.
+ *
+ * Call utfIterator() to have the compiler deduce the UnitIter type.
+ *
+ * For reverse iteration, either use this iterator directly as in <code>*--iter</code>
+ * or wrap it using std::make_reverse_iterator(iter).
+ *
+ * @tparam CP32 Code point type: UChar32 (=int32_t) or char32_t or uint32_t;
+ *              should be signed if UTF_BEHAVIOR_NEGATIVE
+ * @tparam behavior How to handle ill-formed Unicode strings
+ * @tparam UnitIter An iterator (often a pointer) that returns a code unit type:
+ *     UTF-8: char or char8_t or uint8_t;
+ *     UTF-16: char16_t or uint16_t or (on Windows) wchar_t
+ * @draft ICU 78
+ * @see utfIterator
+ */
+template<typename CP32, UTFIllFormedBehavior behavior, typename UnitIter, typename = void>
+class UTFIterator {
+    static_assert(sizeof(CP32) == 4, "CP32 must be a 32-bit type to hold a code point");
+    using Impl = UTFImpl<CP32, behavior, UnitIter>;
+
+    // Proxy type for operator->() (required by LegacyInputIterator)
+    // so that we don't promise always returning CodeUnits.
+    class Proxy {
+    public:
+        explicit Proxy(CodeUnits<CP32, UnitIter> &units) : units_(units) {}
+        CodeUnits<CP32, UnitIter> &operator*() { return units_; }
+        CodeUnits<CP32, UnitIter> *operator->() { return &units_; }
+    private:
+        CodeUnits<CP32, UnitIter> units_;
+    };
+
+public:
+    /** C++ iterator boilerplate @internal */
+    using value_type = CodeUnits<CP32, UnitIter>;
+    /** C++ iterator boilerplate @internal */
+    using reference = value_type;
+    /** C++ iterator boilerplate @internal */
+    using pointer = Proxy;
+    /** C++ iterator boilerplate @internal */
+    using difference_type = typename std::iterator_traits<UnitIter>::difference_type;
+    /** C++ iterator boilerplate @internal */
+    using iterator_category = std::conditional_t<
+        std::is_base_of_v<
+            std::bidirectional_iterator_tag,
+            typename std::iterator_traits<UnitIter>::iterator_category>,
+        std::bidirectional_iterator_tag,
+        std::forward_iterator_tag>;
+
+    // TODO: Should we enable this only for a bidirectional_iterator?
```

Comment:
The documentation seems to answer this TODO in the affirmative, and the implementation in the negative.

I think the documentation is right.

Reply:
Actually the documentation talks about forward_iterator and the TODO about bidirectional_iterator.

Either way, as currently implemented, I think it does not actually require a forward_iterator (pretty sure you could construct it with an istreambuf_iterator), and that seems iffy.

Reply:
This TODO is only about the following three-parameter constructor. For a forward_iterator, providing a start sentinel is unnecessary but harmless.

I am not sure whether enable_if_t even works on a constructor. It resolves to a type (void or second tparam), and a constructor has no return type.

Note that when the UnitIter is only an input_iterator, then the following template specialization is used, and that one does not have this constructor signature.

Note also that I hide that specialization from doxygen, so the API docs above need to speak for both.

Reply:
I think we can resolve this TODO at a later time. I will make a note of it.

Reply:
> Note that when the UnitIter is only an input_iterator, then the following template specialization is used, and that one does not have this constructor signature.

Ah yes, I forgot about that. Nevermind.